semantic_search.py

```diff
@@ -0,0 +1,85 @@
+import argparse
+import os
+from functools import cache
+
+import numpy as np
+import torch
+from sentence_transformers import util
+
+
+@cache
+def get_title_embeddings(filename: str) -> np.ndarray:
+    return np.load(filename + "_title_emb.npy")
+
+
+@cache
+def get_abs_embeddings(filename: str) -> np.ndarray:
+    return np.load(filename + "_abs_emb.npy")
+
+
+@cache
+def get_cat_ohe(filename: str) -> np.ndarray:
+    return np.load(filename + "_cat_ohe.npy")
+
+
+def find_row_numbers(ids: list, arxiv_ids: np.ndarray) -> np.ndarray:
+    return [np.where(arxiv_ids == id)[0][0] for id in ids]
+
+
+def top_hits(
+    title_embeddings: np.ndarray,
+    abs_embeddings: np.ndarray,
+    cat_ohe: np.ndarray,
+    rows: list,
+    n: int = 10,
+) -> np.ndarray:
+    assert (
+        title_embeddings.shape[0] == abs_embeddings.shape[0]
+    ), "Not the same number of titles and abstracts"
+
+    """Top n hits that match a row in embeddings, returns their indices
+    and cosine similarities"""
+    corpus_embeddings = np.concatenate(
+        (title_embeddings, abs_embeddings, cat_ohe), axis=1
+    )
+    corpus_embeddings = torch.from_numpy(corpus_embeddings)
```

Comment:
Idea: what about creating a function that returns the corpus embedding directly instead of building it at prediction time? You can do it either in here, or in the script that generates the embedding (ie, return just one file instead of 3)
You would also be able simplify the row lookup as you won't need to concatenate anything.

Reply:
With the new model, I am only using title + abstract (no categories anymore, don't think it's needed). So, no need to concatenate everything, check the new version to see if there are comments on this